p2p/libp2p/netMessenger_test.go

```diff
@@ -1187,6 +1211,65 @@ func TestLibp2pMessenger_SendDataThrottlerShouldReturnCorrectObject(t *testing.T
 	_ = mes.Close()
 }
 
+func TestLibp2pMessenger_SendDirectShouldNotBroadcastIfMessageIsPartiallyInvalid(t *testing.T) {
+	numOfPeers := 4
+	_, peers := createMockNetworkOfN(numOfPeers)
+	connectPeersFullMesh(peers)
+
+	broadcastMsgResolver := []byte("broadcast resolver msg")
+	directMsgResolver := []byte("resolver msg")
+	msgRequester := []byte("resolver msg is partially valid, mine is ok")
+	mesProcessorRequester := &mock.MessageProcessorStub{
+		ProcessMessageCalled: func(message p2p.MessageP2P, broadcastHandler func(buffToSend []byte)) error {
+			if !bytes.Equal(message.Data(), directMsgResolver) {
+				// pass through all other messages
+				return nil
+			}
+
+			if broadcastHandler != nil {
+				broadcastHandler(msgRequester)
+			}
+
+			return errors.New("resolver msg is partially valid")
+		},
+	}
+
+	mesProcessorResolverAndOtherPeers := &mock.MessageProcessorStub{
+		ProcessMessageCalled: func(message p2p.MessageP2P, _ func(buffToSend []byte)) error {
+			if bytes.Equal(message.Data(), msgRequester) {
+				assert.Fail(t, "other peers should have not received filtered out requester's message")
+			}
+			return nil
+		},
+	}
+
+	idxRequester := 0
+
+	topic := "testTopic"
+	for i := 0; i < numOfPeers; i++ {
+		_ = peers[i].CreateTopic(topic, true)
+		if i == idxRequester {
+			_ = peers[i].RegisterMessageProcessor(topic, mesProcessorRequester)
+		} else {
+			_ = peers[i].RegisterMessageProcessor(topic, mesProcessorResolverAndOtherPeers)
+		}
+	}
+
+	fmt.Println("Delaying for peer connections and topic broadcast...")
+	time.Sleep(time.Second * 5)
+
+	idxResolver := 1
+	fmt.Println("broadcasting a message")
+	peers[idxResolver].Broadcast(topic, broadcastMsgResolver)
+
+	time.Sleep(time.Second)
+
+	fmt.Println("sending a direct message")
+	_ = peers[idxResolver].SendToConnectedPeer(topic, directMsgResolver, peers[idxRequester].ID())
+
+	time.Sleep(time.Second * 2)
```

Comment:
add some final assert check, that the code was reached as many time as you hoped

Reply:
Good catch! It could otherwise give false positive results.

Reply:
done